File: apps/web/src/api/student-dashboard.ts
import { apiClient } from "./client";

export interface KpiItem {
  id: string;
  label: string;
  value: string;
  sub?: string | null;
  delta?: string | null;
  delta_dir?: string | null;
  value_class?: string | null;
}

export interface HealthFactor {
  label: string;
  value: number;
  weight_pct: number;
}

export interface CoachItem {
  title: string;
  why: string;
  cta: string;
  coach_key: string;
}

export interface GrowthStat {
  value: string;
  label: string;
  icon: string;
}

export interface DailyBrief {
  text: string;
  bullets: string[];
}

export interface SemesterTrendItem {
  label: string;
  value_pct: number;
}

export interface SubjectHealthItem {
  course_code: string;
  course_name: string;
  status: string;
  tier: string;
}

export interface StudentDashboard {
  student_id: string;
  name: string;
  first_name: string;
  programme_name: string | null;
  semester_label: string;
  session_label: string;
  day_summary: string;
  health_score: number;
  health_label: string;
  health_factors: HealthFactor[];
  health_narrative: string;
  needs_attention: boolean;
  attention_banner: { title: string; body: string } | null;
  kpis: KpiItem[];
  daily_brief: DailyBrief;
  coach_items: CoachItem[];
  growth_stats: GrowthStat[];
  semester_trend: SemesterTrendItem[];
  subject_health: SubjectHealthItem[];
  cgpa: number | null;
  study_streak_days: number;
  on_time_submissions: number;
}

export interface TimetableSession {
  id: string;
  course_id: string | null;
  course_code: string | null;
  course_name: string | null;
  session_date: string;
  start_time: string;
  end_time: string | null;
  session_type: string;
  title: string;
  room: string | null;
  faculty_name: string | null;
  notes: string | null;
  status: string;
}

export interface TimetableDay {
  date: string;
  summary: string;
  sessions: TimetableSession[];
}

export interface Assignment {
  id: string;
  course_id: string;
  course_code: string | null;
  course_name: string | null;
  title: string;
  due_at: string;
  due_label: string;
  status: string;
  progress_pct: number;
  priority: string;
}

export interface AssignmentsList {
  open_count: number;
  items: Assignment[];
}

export interface AttendanceCourseDetail {
  course_id: string;
  course_code: string;
  course_name: string;
  percentage: number;
  present_sessions: number;
  total_sessions: number;
  below_threshold: boolean;
}

export interface AttendanceDetail {
  overall_pct: number;
  required_pct: number;
  predicted_pct: number | null;
  margin_sessions: number | null;
  note: string;
  courses: AttendanceCourseDetail[];
}

export interface ExamSubjectReadiness {
  course_id: string;
  course_code: string;
  course_name: string;
  readiness_pct: number;
  exam_name: string | null;
  exam_date: string | null;
  days_until_exam: number | null;
}

export interface ExamPrep {
  overall_readiness: number;
  headline: string;
  tip: string;
  subjects: ExamSubjectReadiness[];
}

export interface CareerOpportunity {
  title: string;
  subtitle: string;
  icon: string;
}

export interface CareerProfile {
  readiness_score: number;
  skills: string[];
  opportunities: CareerOpportunity[];
  credits_completed: number;
  credits_required: number;
  narrative: string;
}

export interface StudentNotification {
  id: string;
  title: string;
  body: string | null;
  tone: string;
  created_at: string;
  read_at: string | null;
  time_label: string;
}

export interface StudentActivity {
  id: string;
  activity_type: string;
  summary: string;
  created_at: string;
  time_label: string;
}

export interface FeesDetail {
  total_due: string;
  total_paid: string;
  total_balance: string;
  overdue_count: number;
  note: string;
  items: {
    term: string;
    fee_head: string;
    amount_due: string | null;
    amount_paid: string | null;
    due_date: string | null;
    status: string | null;
    balance: string | null;
  }[];
}

export interface CampusAnnouncement {
  id: string;
  title: string;
  body: string | null;
  location: string | null;
  published_at: string;
  closes_at: string | null;
  time_label: string;
}

export async function getStudentDashboard(studentId: string): Promise<StudentDashboard> {
  const { data } = await apiClient.get<StudentDashboard>(`/students/${studentId}/dashboard`);
  return data;
}

export async function getTimetable(studentId: string, date?: string): Promise<TimetableDay> {
  const { data } = await apiClient.get<TimetableDay>(`/students/${studentId}/timetable`, {
    params: date ? { date } : undefined,
  });
  return data;
}

export async function getAssignments(studentId: string, status?: string): Promise<AssignmentsList> {
  const { data } = await apiClient.get<AssignmentsList>(`/students/${studentId}/assignments`, {
    params: status ? { status } : undefined,
  });
  return data;
}

export async function getAttendanceDetail(studentId: string): Promise<AttendanceDetail> {
  const { data } = await apiClient.get<AttendanceDetail>(`/students/${studentId}/attendance`);
  return data;
}

export async function getFeesDetail(studentId: string): Promise<FeesDetail> {
  const { data } = await apiClient.get<FeesDetail>(`/students/${studentId}/fees`);
  return data;
}

export async function getExamPrep(studentId: string): Promise<ExamPrep> {
  const { data } = await apiClient.get<ExamPrep>(`/students/${studentId}/exam-prep`);
  return data;
}

export async function getCareerProfile(studentId: string): Promise<CareerProfile> {
  const { data } = await apiClient.get<CareerProfile>(`/students/${studentId}/career`);
  return data;
}

export async function getStudentNotifications(studentId: string): Promise<StudentNotification[]> {
  const { data } = await apiClient.get<StudentNotification[]>(`/students/${studentId}/notifications`);
  return data;
}

export async function getStudentActivity(studentId: string): Promise<StudentActivity[]> {
  const { data } = await apiClient.get<StudentActivity[]>(`/students/${studentId}/activity`);
  return data;
}

export async function getCampusAnnouncements(): Promise<CampusAnnouncement[]> {
  const { data } = await apiClient.get<CampusAnnouncement[]>("/campus/announcements");
  return data;
}

File: apps/api/app/schemas/student_dashboard.py
from datetime import date, datetime, time
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class CourseRef(BaseModel):
    id: UUID
    code: str
    name: str


class AttendanceSummary(BaseModel):
    course_id: UUID
    course_code: str | None = None
    course_name: str | None = None
    total_sessions: int
    present_sessions: int
    percentage: float


class InternalMarkSummary(BaseModel):
    course_id: UUID | None
    course_code: str | None = None
    course_name: str | None = None
    assessment_type: str
    attempt: int
    max_marks: Decimal
    obtained: Decimal


class FeeSummary(BaseModel):
    term: str
    fee_head: str
    amount_due: Decimal | None
    amount_paid: Decimal | None
    status: str | None


class FeeLineItem(BaseModel):
    term: str
    fee_head: str
    amount_due: Decimal | None
    amount_paid: Decimal | None
    due_date: date | None
    status: str | None
    balance: Decimal | None


class FeesDetailResponse(BaseModel):
    total_due: Decimal
    total_paid: Decimal
    total_balance: Decimal
    overdue_count: int
    note: str
    items: list[FeeLineItem]


class Student360Response(BaseModel):
    id: UUID
    canonical_roll_no: str
    name: str
    dob: date | None
    gender: str | None
    category: str | None
    email: str | None
    phone: str | None
    admission_year: int | None
    programme_id: UUID | None
    programme_name: str | None = None
    status: str | None
    attendance_summary: list[AttendanceSummary]
    marks: list[InternalMarkSummary]
    fees: list[FeeSummary]
    course_lookup: dict[str, CourseRef] = Field(default_factory=dict)


class TimetableSessionResponse(BaseModel):
    id: UUID
    course_id: UUID | None
    course_code: str | None = None
    course_name: str | None = None
    session_date: date
    start_time: time
    end_time: time | None
    session_type: str
    title: str
    room: str | None
    faculty_name: str | None
    notes: str | None
    status: str  # done | now | next | free | urgent


class TimetableDayResponse(BaseModel):
    date: date
    summary: str
    sessions: list[TimetableSessionResponse]


class AssignmentResponse(BaseModel):
    id: UUID
    course_id: UUID
    course_code: str | None = None
    course_name: str | None = None
    title: str
    due_at: datetime
    due_label: str
    status: str
    progress_pct: int
    priority: str


class AssignmentsListResponse(BaseModel):
    open_count: int
    items: list[AssignmentResponse]


class AttendanceCourseDetail(BaseModel):
    course_id: UUID
    course_code: str
    course_name: str
    percentage: float
    present_sessions: int
    total_sessions: int
    below_threshold: bool


class AttendanceDetailResponse(BaseModel):
    overall_pct: float
    required_pct: float
    predicted_pct: float | None
    margin_sessions: int | None
    note: str
    courses: list[AttendanceCourseDetail]


class ExamSubjectReadiness(BaseModel):
    course_id: UUID
    course_code: str
    course_name: str
    readiness_pct: int
    exam_name: str | None = None
    exam_date: date | None = None
    days_until_exam: int | None = None


class ExamPrepResponse(BaseModel):
    overall_readiness: int
    headline: str
    tip: str
    subjects: list[ExamSubjectReadiness]


class CareerOpportunity(BaseModel):
    title: str
    subtitle: str
    icon: str = "briefcase"


class CareerProfileResponse(BaseModel):
    readiness_score: int
    skills: list[str]
    opportunities: list[CareerOpportunity]
    credits_completed: int
    credits_required: int
    narrative: str


class StudentNotificationResponse(BaseModel):
    id: UUID
    title: str
    body: str | None
    tone: str
    created_at: datetime
    read_at: datetime | None
    time_label: str


class StudentActivityResponse(BaseModel):
    id: UUID
    activity_type: str
    summary: str
    created_at: datetime
    time_label: str


class CampusAnnouncementResponse(BaseModel):
    id: UUID
    title: str
    body: str | None
    location: str | None
    published_at: datetime
    closes_at: datetime | None
    time_label: str


class KpiItem(BaseModel):
    id: str
    label: str
    value: str
    sub: str | None = None
    delta: str | None = None
    delta_dir: str | None = None  # up | down
    value_class: str | None = None


class HealthFactor(BaseModel):
    label: str
    value: int
    weight_pct: int


class CoachItem(BaseModel):
    title: str
    why: str
    cta: str
    coach_key: str


class GrowthStat(BaseModel):
    value: str
    label: str
    icon: str


class DailyBriefResponse(BaseModel):
    text: str
    bullets: list[str]


class SemesterTrendItem(BaseModel):
    label: str
    value_pct: float


class SubjectHealthItem(BaseModel):
    course_code: str
    course_name: str
    status: str  # strong | healthy | watch
    tier: str  # low | watch | high — internal mapping only


class StudentDashboardResponse(BaseModel):
    student_id: UUID
    name: str
    first_name: str
    programme_name: str | None
    semester_label: str
    session_label: str
    day_summary: str
    health_score: int
    health_label: str  # On track | Needs attention | Building momentum
    health_factors: list[HealthFactor]
    health_narrative: str
    needs_attention: bool
    attention_banner: dict[str, str] | None = None
    kpis: list[KpiItem]
    daily_brief: DailyBriefResponse
    coach_items: list[CoachItem]
    growth_stats: list[GrowthStat]
    semester_trend: list[SemesterTrendItem]
    subject_health: list[SubjectHealthItem]
    cgpa: float | None
    study_streak_days: int
    on_time_submissions: int
